uring/src/main/scala/fs2/io/uring/net/UringDatagramSocket.scala

```diff
@@ -35,7 +36,45 @@ private[net] final class UringDatagramSocket[F[_]](
   private[this] def recvmsg(msg: Ptr[msghdr], flags: Int): F[Int] =
     ring.call(io_uring_prep_recvmsg(_, fd, msg, flags))
 
-  def read: F[Datagram] = ???
+  def read: F[Datagram] =
+    readMutex.lock.surround {
+      for {
+        buf <- buffer.get(defaultReadSize)
+
+        iov <- F.delay {
+          val iov = stackalloc[uio.iovec]()
+          iov._1 = buf
+          iov._2 = defaultReadSize.toULong
+          iov
+        }
+
+        remoteAddr <- F.delay {
+          stackalloc[sockaddr_storage]()
+        }
+
+        msg <- F.delay {
+          val msg = stackalloc[msghdr]()
+          msg._1 = remoteAddr.asInstanceOf[Ptr[Byte]]
+          msg._2 = sizeof[sockaddr_storage].toUInt
+          msg._3 = iov
+          msg._4 = 1
+          msg._5 = null
+          msg._6 = 0.toUInt
+          msg._7 = 0
```

Comment:
There is an ops class you can import so you can call the members by their name, instead of `_1`, `_2`, etc. :) same for `iovec` too I think.

https://github.com/scala-native/scala-native/blob/c98eeeba8fa757f204b1730c2a0bca11f9bae4de/posixlib/src/main/scala/scala/scalanative/posix/sys/socket.scala#L455

Reply:
Great! it makes the code more readable :) I found the iovecOps but in the project version is missing I think.

https://github.com/scala-native/scala-native/blob/c98eeeba8fa757f204b1730c2a0bca11f9bae4de/posixlib/src/main/scala/scala/scalanative/posix/sys/uio.scala